### DISCUSSION TITLE
Unique index creation on mssql

### DISCUSSION THREAD

**Discussion by fairking** [fluentmigrator/fluentmigrator#1520] [General]
I think the following line:
```
Create.Index("uq_users_email").OnTable("users").OnColumn("email").Unique();
```
must execute the following sql query on the mssql db:
```
CREATE UNIQUE NONCLUSTERED INDEX uq_users_email ON users(email) WHERE email IS NOT NULL;
```
Regarding the issue: https://stackoverflow.com/questions/767657/how-do-i-create-a-unique-constraint-that-also-allows-nulls

I think having the migrator it should end up with the same result on all databases.

Otherwise with many databases we are forced to do the following hack:
```
IfDatabase("oracle", "postgres", "sqlite").Create.Index("uq_users_email").OnTable("users").OnColumn("email").Unique();
IfDatabase("sqlserver").Execute.Sql("CREATE UNIQUE NONCLUSTERED INDEX uq_users_email ON users(email) WHERE email IS NOT NULL;");
```

What do you thinkg guys?

**Comment by jzabroski** [comment]
If I understand your point, it's that Microsoft SQL Server unique indexes have funky support for nullable columns.

The actual workaround would be:

```sql
IfDatabase("oracle", "postgres", "sqlite").Create.Index("uq_users_email").OnTable("users").OnColumn("email").Unique();
IfDatabase("mssql").Create.Index("uq_users_email").OnTable("users").OnColumn("email").Unique().Filter("email IS NOT NULL");
```

I need to think for a bit if this is a hack, or if it's a feature that you are forced to do this in this scenario.  I think they're not entirely the same semantics, but I get your point that it stinks that it will fail on mssql processor.  I used to have the connect item for this back when Microsoft Connect was how we tracked issues, but Microsoft deleted Connect and all associated issues.  That Connect item did outline the exact semantic differences, which I am fuzzy on at the moment.

**Reply by fairking** [reply]
> I think they're not entirely the same semantics, but I get your point that it stinks that it will fail on mssql processor

Yes, I mean when we make an unique column I think it should behave in the same way on any database otherwise everyone needs to remember to write an extra hack.

**Reply by jzabroski** [reply]
Yes, but the problem is, generating the right SQL for SQL Server requires knowing the nullability of the column, which we can't always know (unless we query the database, which currently we don't do).  In this case, if you run it on MSSQL, you will discover there is platform-specific semantics, which your developers may want to think about.  Hence my point that if it isn't 100% semantically equivalent, we shouldn't abstract over important implementation details.

@fubar-coder Thoughts?

**Reply by fairking** [reply]
> SQL Server requires knowing the nullability of the column

It could be not necessery as the index check `WHERE email IS NOT NULL` doesn't throw any error on non-nullable columns. However it can slow the performance.

> Hence my point that if it isn't 100% semantically equivalent, we shouldn't abstract over important implementation details

`services.AddFluentMigratorCore(semantic: true)` or `migratorBuilder.AddSqlServer(semantic: true);` or `.Unique(semantic: true)`.

Also it depends how often unique indexes are used by developers. If it's a rare case then it might not worthy. But I guess there can be many more semantic cases across different databases, not just indexes and not just mssql (eg. datatypes).

Whatever is possible the main goal to make things simple must remain 👍
